### ISSUE TITLE
Create permanent identifiers for alleles

### ISSUE THREAD

**Issue by kimrutherford** [pombase/pombase-chado#986]
See also:
  - pombase/pombase-chado#840
  - https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2001414

**Comment by ValWood** [comment]
OK great. Don't hurry,  I can say in progress if we mention it at all.

**Comment by kimrutherford** [comment]
> Once that's done, it will be straightforward to write the current allele IDs, names etc. to a file.

Turns out that writing the allele details to a file was straightforward.  The allele information was already collected in the correct format in memory during the Chado-to-website processing step.  It was just a few lines of extra code to get useful file: https://curation.pombase.org/dumps/latest_build/misc/allele_details.json

**Comment by kimrutherford** [comment]
> I've had a think and this might not be too tricky to implement once this issue is done:
>    - pombase/website#1294

I've thought some more.  It doesn't make sense to have allele pages until we have permanent identifiers so these two issues need to be completed in parallel.

**Comment by kimrutherford** [comment]
> The idea would be that once an allele had an ID, it would always be passed along to next night's load. We would be initialising Chado with the previous alleles before we start loading from Canto and PHAF files.

I think this is mostly solved.  I've written a new loading step that reads the allele IDs and details from [JSON file from the previous load](https://github.com/pombase/pombase-chado/issues/986#issuecomment-1158615560).  Since the ID and details will be pre-loaded, the allele IDs should stay the same from load to load.

I'm going to wait until Friday night before I activate the new step in case something goes wrong.  My plan then is to check the load and compare with Thursday night's to make sure it worked OK.

**Comment by kimrutherford** [comment]
> I'm going to wait until Friday night before I activate the new step in case something goes wrong.

This is on hold until I fix: pombase/pombase-chado#992

**Comment by kimrutherford** [comment]
Note to self: don't forgot to change the JaponicusDB build script to match any pombe changes.

**Comment by kimrutherford** [comment]
(From: https://github.com/pombase/website/issues/1294#issuecomment-1218509827)

We need to think about what happens when an allele is renamed, or its description changes.  We talked on Zoom about implementing a new tool (maybe within Canto) that will allow alleles to be renamed in all sessions at once.  I've made an issue:
  - pombase/canto#2669

We also need to think about what happens when an allele is deleted but there is a plan for that: Chado has a "is_obsolete" field for each feature.  If we set that to true for deleted alleles I think everything will work out.

**Comment by kimrutherford** [comment]
Now waiting on:

   - pombase/canto#2758
   - pombase/pombase-chado#1107

**Comment by kimrutherford** [comment]
I've done some test runs today and it's all in a better state than I remember.  :-)

It will probably take quite a few nightly loads to get things right so I suggest that we disable Canto and public pombase.org updates on Thursday night.  That will give me Friday to commit and check the code and config changes.  And then I'll have the weekend to run lots of test loads until things are working.

Does that sounds OK?

**Comment by ValWood** [comment]
Yes. I still need to do some fixes from the logs. Which log file was it?
I sent myself a link, but it's to the wrong thing.

**Comment by kimrutherford** [comment]
I think this is the highest priority: https://curation.pombase.org/dumps/builds/pombase-build-2024-02-19/logs/log.2024-02-19-00-56-47.chado_checks.duplicated_allele_names

Could you look at this too when you get a chance?: https://github.com/pombase/curation/issues/3561

I'll double check the other log files before Friday to see if there is anything else that needs fixing urgently.

**Comment by ValWood** [comment]
I will do these today

**Comment by ValWood** [comment]
> Does that sounds OK?
 
and yes!

**Comment by ValWood** [comment]
These should be fixed for tomorrow

https://curation.pombase.org/dumps/builds/pombase-build-2024-02-19/logs/log.2024-02-19-00-56-47.chado_checks.duplicated_allele_names

**Comment by kimrutherford** [comment]
> These should be fixed for tomorrow

Thanks!

**Comment by kimrutherford** [comment]
Hi Val.

There is still a duplicate allele name:

```
Check for two or more alleles with the same name - CHECK FAILURE:  expected 0 but got 2
name	uniquename	description	allele_type	canto_session
  prp3-3	SPAC29E6.02:allele-5	unknown	unknown	e8547aef6b97c8ef
  prp3-3	SPAC3A12.11c:allele-3	unknown	unknown	7d409d497eb075ca
```

Over the weekend I did a couple of full test loads on my desktop with the new Canto code and the new loading code.  Everything seemed to work as planned.

**Comment by kimrutherford** [comment]
Checklist for putting the new allele system into production:
  - [ ] fix all allele warnings from the Canto load log
  - [ ] export the everything in the Canto database, including unapproved sessions, then try to load it into Chado
       - check the log and fix any other allele warnings we can
  - [ ] run a normal load to get a set of allele uniquenames in `allele_summaries.json`
  - [ ] merge [Canto allele branch](https://github.com/pombase/canto/tree/issue-2758-disable-edits-for-existing-alleles)
     - [ ] make a backup 
     - [ ] update Canto DB schema
     - [ ] consider running `etc/update_canto_alleles_from_chado.pl`
     - [ ] run `etc/set_ids_from_chado.pl`
     - [ ] put the main Canto in admin only mode for testing (needs implementing)
     - [ ] test Canto changes
  - [ ] merge [website branch](https://github.com/pombase/website/tree/issue-1255-gene-alleles-page-fixes) / [PR](https://github.com/pombase/website/pull/2149)
  - [ ] un-comment command at top of nightly script that loads the previous day's alleles (look for "loading alleles from previous load")
  - [ ] Check Chado
  - [ ] test Canto, especially existing alleles
  - [ ] test website

**Comment by kimrutherford** [comment]
Hi Val.

There are some Canto load errors that might be easier to fix before the allele systematic ID changes.  Are these hard to fix?

https://curation.pombase.org/dumps/builds/pombase-build-2024-02-26/logs/log.2024-02-25-22-16-33.curation-tool-data-load-output

**Comment by ValWood** [comment]
On my to do list for this week.

**Comment by ValWood** [comment]
I think prp3-1 is an allele of cwf5 but need to double check this

<img width="599" alt="Screenshot 2024-02-26 at 12 12 39" src="https://github.com/pombase/pombase-chado/assets/7359272/86af270c-e485-4934-adec-391ded0cd0d6">

**Comment by ValWood** [comment]
fixed prp3-1

**Comment by ValWood** [comment]
I think I have eventually cleared this log...will check tomorrow.....

**Comment by ValWood** [comment]
another go, hopefully tomorrow...

**Comment by ValWood** [comment]
it worked, but I have new ones, will fix today!

**Comment by kimrutherford** [comment]
I've now merged this code into the master branch (locally).  It was a bit painful because there were conflicts with the code changes from pombase/canto#2544.

I've also merged the changes into the test Canto so we can test things: https://curation.pombase.org/test/curs/4666975359de04dd/genotype_manage


Note to self: branch issue-2758-disable-edits-for-existing-alleles-merged

**Comment by kimrutherford** [comment]
> I've now merged this code into the master branch (locally).

Sorry, that wasn't clear.  I've merged the code for preventing allele changes in Canto when you select an existing allele: pombase/canto#2758

**Comment by ValWood** [comment]
We found that we could not  edit any annotation for an existing allele in *any *
annotation where the **allele** was used in a genetic interaction, even if that **annotation** was not used in the interaction.

e.g.
<img width="1403" alt="Screenshot 2024-03-18 at 11 36 50" src="https://github.com/pombase/pombase-chado/assets/7359272/7b3d607e-129a-4ea1-8130-5e8f23e67228">

Let me know if you want to discuss.

CC @PCarme

**Comment by ValWood** [comment]
For example in the same session I want to change this evidence code to microscopy but I can't. I can't do it by copy/edit delete either because delete is also greyed out

<img width="1306" alt="Screenshot 2024-03-18 at 11 37 58" src="https://github.com/pombase/pombase-chado/assets/7359272/779301cb-cfd4-422d-afe3-b47ad9712e3e">

**Comment by kimrutherford** [comment]
> We found that we could not edit any annotation for an existing allele in *any * annotation

I don't think this is due to the allele identifier changes.  It's because of: pombase/canto#2740

I've pasted your comment into that issue.

**Comment by kimrutherford** [comment]
I started work on this again yesterday.  First step was to get the branch with Canto changes updated:
  - https://github.com/pombase/canto/tree/issue-986-permanent-allele-identifiers

I did a bit of testing and fixed some bugs but I'm sure there are more.  Most of the changes are to the allele dialog and the code behind it.

Next step is to re-test the script that updates the Canto database with the allele identifiers from Chado.

Here is the project tracker:
 - https://github.com/orgs/pombase/projects/2

**Comment by kimrutherford** [comment]
Checklist for putting the new allele system into production (updated):
  - [ ] update the Canto schema to v37 which includes the new `external_uniquename` on alleles
    - this can be done anytime 
  - [ ] pause new sessions and curation
  - [ ] approve as many sessions as possible
     - this will reduce the chance of an allele in Canto missing out on getting the correct unique ID from Chado
  - [ ] fix all allele warnings from the Canto load log
  - [ ] fix duplicate description warnings in log file: `*.chado_checks.duplicate_allele_descriptions`
  - [ ] export the everything in the Canto database, including unapproved sessions, then try to load it into Chado as a test run
       - check the log and fix any other allele warnings we can
  - [ ] run a normal load to get a set of allele uniquenames in `pombe-embl/supporting_files/nightly_load_results/allele_summaries.json`
  - [ ] un-comment (and git commit) command at top of nightly script that loads the previous day's alleles (look for "loading alleles from previous load")
  - [ ] merge [Canto allele branch](https://github.com/pombase/canto/tree/issue-986-permanent-allele-identifiers)
  - [ ] make a Canto backup 
  - [ ] update Canto DB schema
  - [ ] consider running `etc/update_canto_alleles_from_chado.pl`
  - [ ] run `etc/set_ids_from_chado.pl` (see pombase/canto#2770)
  - [ ] put the main Canto in admin only mode for testing (needs implementing)
  - [ ] test Canto changes
  - [ ] export sessions from Canto, which now include `external_uniquename` for alleles
  - [ ] do another Chado test load
  - [ ] check Chado
    - [ ] do the allele IDs remain the same? 
    - [ ] is the `is_obsolete` flag on alleles set to true?  (See https://github.com/pombase/pombase-chado-json/commit/d8fb3d288b9c1d73d2f0be9b2ed8e627b51e7be4) 
  - [ ] test Canto, especially existing alleles
  - [ ] merge [website branch](https://github.com/pombase/website/tree/issue-1255-gene-alleles-page-fixes) / [PR](https://github.com/pombase/website/pull/2149)
  - [ ] test website
  - [ ] add `etc/set_ids_from_chado.pl` to the nightly Canto cron job (maybe - the changes to the Canto allele edit dialog should make this unnecessary)

**Comment by kimrutherford** [comment]
I've done a full test load on my desktop with the updated code.  Everything seems OK including my local version of Canto ([this change for example](https://github.com/pombase/canto/issues/2758)).
I'm sure there will be more issues when we attempt to switch over but it's all looking positive.

We now need a few days of calm to go through this process:
  - https://github.com/pombase/pombase-chado/issues/986#issuecomment-4316916670

**Comment by ValWood** [comment]
not yet then!

**Comment by kimrutherford** [comment]
It would make life a little easier if some of these can be fixed before we try to add systematic IDs:

 - https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.AlleleNotStartingWithGeneName  
 - https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.duplicate_allele_descriptions
 - https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.unknown_allele_name_synonym_matches

Thanks.

**Comment by ValWood** [comment]
I'm currfently working through this file:
https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.AlleleNotStartingWithGeneName

@kimrutherford  could you replace 
SPAC31G5.10delta1759-2859	eta2	SPAC31G5.10	
SPAC24B11.05::Tn-seq	ifn1	SPAC24B11.05	
SPCC63.06::Tn-seq	wdr89	SPCC63.06	
SPBC28E12.04::Tn-seq	cbp7	SPBC28E12.04	
SPAC144.05::Tn-seq	irc20	SPAC144.05	
SPAP8A3.07c::Tn-seq	aro3	SPAP8A3.07c	
SPBC577.03c::Tn-seq	nat9	SPBC577.03c	
SPBC11C11.05::Tn-seq	knh1	SPBC11C11.05
witht he gene name in the appropriate gaf,

@PCarme  could you look at these 2, I think they'll be easier for you 
mat1-Mm143	mat1-Mi	SPBC23G7.17c	8fa55d6ed8dab65f
P-factorDeltaLeu	map2	SPCC1795.06	4acb2f022c79e2b7

I'll do the others

**Comment by kimrutherford** [comment]
> could you replace

Yep, no problem.

**Comment by kimrutherford** [comment]
"SPAC31G5.10delta1759-2859" seems a bit odd.  Should the name just be "eta2delta1759-2859"?

**Comment by ValWood** [comment]
Should the name just be "eta2delta1759-2859"

yes, sorry, i didn't explain properly

**Comment by kimrutherford** [comment]
OK, thanks.  Those fixes are done.

**Comment by PCarme** [comment]
> [@PCarme](https://github.com/PCarme) could you look at these 2, I think they'll be easier for you
> mat1-Mm143 mat1-Mi SPBC23G7.17c 8fa55d6ed8dab65f
> P-factorDeltaLeu map2 SPCC1795.06 4acb2f022c79e2b7

These 2 should be fixed now.

**Comment by ValWood** [comment]
[AlleleNotStartingWithGeneName](https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.AlleleNotStartingWithGeneName)
should be empty tomorrow

**Comment by ValWood** [comment]
[chado_checks.duplicate_allele_descriptions](https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.duplicate_allele_descriptions)
should be empty tomorrow

**Comment by ValWood** [comment]
@kimrutherford  could you locate the source (PMID)  of 
  plo1-ts4	SPAC23C11.16	R132Q	synonym	b687ff7ab68d1820,5f14bab21adf5958,a2468b284d810cf8
  plo1-1	SPAC23C11.16	R132Q	synonym	

plo1-1 for me? thanks,

in
https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.unknown_allele_name_synonym_matches

**Comment by kimrutherford** [comment]
plo1-ts4:
```
5f14bab21adf5958  PMID:40161431
a2468b284d810cf8  PMID:12546793
b687ff7ab68d1820  PMID:11250892
```

plo1-1:
```
28f15204b6b233af  PMID:15537703
5f14bab21adf5958  PMID:40161431
67631dafe77debd1  PMID:9852154
6a5d03f63d343148  PMID:19417002
977c41e138dc2e0d  PMID:10930468
a765c0082a3d7619  PMID:19075108
d4228b03e96911f7  PMID:15857958
e68d23abf86a3c7c  PMID:34674264
```

**Comment by ValWood** [comment]
ah OK, I didn't see that because of the line wrapping. I'll align on plo1-1 making  plo1-ts4 a synonym

**Comment by ValWood** [comment]
I want to change plo1-ts4 in b687ff7ab68d1820 to plo1-1 and make plots4 a synonym, but  Canto won't let me do that, it will only add plo1-1 as unknown

**Comment by kimrutherford** [comment]
What's the message?

**Comment by ValWood** [comment]
Actually I think it will be OK. I change to plo1-1 with plo1-ts4 as synonym, and I think the loading will figure out the rest...

**Comment by ValWood** [comment]
All should  be done tomorrow except 4 cdc15 alleles where I am waiting for confirmation

**Comment by ValWood** [comment]
wow, this is a long ticket.

Kim do you want to make a new ticket with the outstanding tasks?

My final bit:

For this one 
cdc15-127	SPAC20G8.05c	S50P	synonym	1c9a3835487bb72e,f1b9bb385a8edc45,2598fac4f5efa38a,de3fe3d88f14e4a8
  cdc15-287	SPAC20G8.05c	S50P	synonym	f1b9bb385a8edc45,1eaa4ef71c675ac0,19eeb17f636dc21f,7ced7ee21306a509,482d360415391124

Kathy would like the primary name of  cdc15-287 as that came first.
could you automate the replacements because they will  probably be used in multiple multi alleles.

I will fix the other one.

The rest I/Pascal  fixed are still present in the logs today, but I guess they will disappear soon?
v

**Comment by ValWood** [comment]
i'm confused why the ones i fixed in canto still arent showing as  fixed here:

https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.AlleleNotStartingWithGeneName
https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.duplicate_allele_descriptions
https://curation.pombase.org/dumps/builds/pombase-build-2026-09-17/logs/log.2026-09-16-20-29-06.chado_checks.unknown_allele_name_synonym_matches

**Comment by kimrutherford** [comment]
Those are old log files.  Here are today's:

 - https://curation.pombase.org/dumps/builds/pombase-build-2026-09-21/logs/log.2026-09-20-20-29-04.chado_checks.AlleleNotStartingWithGeneName
 - https://curation.pombase.org/dumps/builds/pombase-build-2026-09-21/logs/log.2026-09-20-20-29-04.chado_checks.duplicate_allele_descriptions
 - https://curation.pombase.org/dumps/builds/pombase-build-2026-09-21/logs/log.2026-09-20-20-29-04.chado_checks.unknown_allele_name_synonym_matches

**Comment by kimrutherford** [comment]
> Kim do you want to make a new ticket with the outstanding tasks?

New issue:
 - pombase/pombase-chado#1377
